tests/virt_autotest/login_console.pm

```diff
@@ -37,6 +37,32 @@ sub login_to_console {
         }
     }
     else {
+        save_screenshot;
+        #offline upgrade requirs upgrading ofline during reboot while online doesn't
+        if (check_var('offline_upgrade', 'yes')) {
+            #boot to upgrade menuentry
+            send_key 'down';
+            send_key 'ret';
+            #wait sshd up
+            assert_screen('sshd-server-started', 180);
+            save_screenshot;
+            #switch to ssh console
+            use_ssh_serial_console;
+            save_screenshot;
+            #start upgrade
+            type_string("DISPLAY= yast.ssh\n");
+            save_screenshot;
+            #wait upgrade finish
+            assert_screen('rebootnow', 2700);
+            save_screenshot;
+            send_key 'ret';
+            #leave ssh console and switch to sol console
+            switch_from_ssh_to_sol_console 'with_reset_consoles';
+            #grub may not showup after upgrade because default GRUB_TERMINAL setting
+            #when fixed in separate PR, will uncomment following line
+            #assert_screen([qw(grub2 grub1)], 120);
+        }
```

Comment:
This whole block looks like more and more special and different then what is used for other test approaches. I have the feeling that maintaining the tests is not improved by this. Can we find a way to reuse more existing test code rather then writing more "virt_autotest" specific one?

Reply:
I am sorry, no specific code. But as you want us to also support tests on IPMI to not break your tests we need to use more common code and not diverge more.

Reply:
Do not worry too much here. 

We hope your group can help take care of xen/kvm host installation on ipmi only, which is the fundamental part of other virtualization tests. But ourselves will take care of all others like guest install/upgrade/migration and so on. 

So we will maintain this piece of code. It seems special, but actually it is just regular offline migration which happens during reboot. The real special part is that it is more complicated to switch between sol and root-ssh consoles , that is on sol console -> type reboot -> upgrade menuentry is booted -> wait sshd ready , then switch to root-ssh console to do the offline upgrade(this is the key idea to make upgrade stable on ipmi, that use ipmi as less as possible), then when rebootnow tag show up, switch from root-ssh console to sol console for regular login.